docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,438 @@
+
+## Motivation
+- A singular provider that combines infrastructure provisioning and Kubernetes bootstrapping is too restrictive. Many current v1alpha1 providers target single cloud providers. If a user wants to install a specific Kubernetes distribution (Kubernetes, OpenShift, Rancher, etc.), it is left to the provider to add support for it. In practical terms, we are not able to guarantee that a user can install the same distribution regardless of infrastructure environment.
+- The Machine custom resource contains fields for provider-specific configuration and status. These are represented as opaque [RawExtensions](https://godoc.org/k8s.io/apimachinery/pkg/runtime#RawExtension), meaning their contents are not directly processed as part of the Machine’s validation.
+- The strict relationship between the Machine controller with provider-specific Actuator doesn't allow reconciling machines with providerSpec of different cloud providers in the same cluster (i.e. hybrid cluster). Separation of provider-specific bits from the Machine controller increases flexibility of the cluster api plane and moves the responsibility for their interpretation and reconciliation into dedicated controllers.
+- Both users and developers are confused by the code organization split. The Cluster API repository currently offers controllers for Cluster, Machine, MachineSet, MachineDeployment, Node. Some of these controllers are meant to be vendored and run as part of a provider's Manager; while some others are generic (MachineDeployment, MachineSet, Node) and meant to be run as part of the Cluster API Manager. As outlined in this proposal, we'd like to streamline the relationship between where the code lives and where it runs. In particular, we'd like to make the Machine type and its controller generic, which will allow for it to be part of the Cluster API Manager.
+
+### Goals
+1. To use Kubernetes Controllers to manage the lifecycle of Machines.
+1. To make bootstrap implementations reusable across any infrastructure provider.
+1. To support pre-boot (e.g. cloud-init, bash scripts, or similar) bootstrapping.
+1. To support bootstrapping different Kubernetes distributions on the same infrastructure provider.
+1. To validate provider-specific content as early as possible (create, update).
+
+### Non-Goals/Future Work
+1. To modify the Cluster object.
+1. To fully implement machine state lifecycle hooks. This must be part of a future proposal that builds on these proposed changes.
+1. To setup OS configuration, install software or any other features related to image building.
+1. To support post-boot configuration of Machine.
+1. To revisit MachineClass role and functionality.
+
```

Comment:
To customize the image beyond the settings required to run kubelet, that is an implementation constraint of the bootstrap provider.

/cc @dhellmann

Reply:
```suggestion
1. To customize the image beyond the settings required to run kubelet, that is an implementation constraint of the bootstrap provider.

```